File: OJ practices/LeetCode题目集/3. Longest Substring Without Repeating Characters(medium).cpp
/*
 * @Autor: SourDumplings
 * @Date: 2019-12-09 14:39:37
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        string thisStr;
        int res = 0;
        for (auto &&c : s)
        {
            auto lastOccur = find(thisStr.begin(), thisStr.end(), c);
            if (lastOccur != thisStr.end())
            {
                if (thisStr.length() > res)
                {
                    res = thisStr.length();
                }
                thisStr.erase(thisStr.begin(), lastOccur + 1);
            }
            thisStr.push_back(c);
        }
        if (thisStr.length() > res)
        {
            res = thisStr.length();
        }
        return res;
    }
};

File: OJ practices/LeetCode题目集/3. Longest Substring Without Repeating Characters(medium).java
/*
 * @Autor: SourDumplings
 * @Date: 2019-12-09 14:21:59
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */

class Solution
{
    public int lengthOfLongestSubstring(String s)
    {
        int res = 0;
        StringBuilder thisStr = new StringBuilder();
        int l = s.length();
        for (int i = 0; i < l; i++)
        {
            String c = s.substring(i, i + 1);
            int lastOccur = thisStr.indexOf(c);
            if (lastOccur != -1)
            {
                if (thisStr.toString().length() > res)
                {
                    res = thisStr.toString().length();
                }
                thisStr = thisStr.delete(0, lastOccur + 1);
            }
            thisStr.append(c);
        }
        if (thisStr.toString().length() > res)
        {
            res = thisStr.toString().length();
        }
        return res;
    }
}